Features/task 2

milan009 opened the pull request.

Commit:
KA-184 Add IRepository interface

Commit:
KA-184 Add ListItemRepository implementation

Commit:
KA-184 Add DI contructors

Commit:
KA-184 Install NSubstitute package

Commit:
KA-184 Add tests that check calls to repository

Commit:
KA-184 Code cleanup

Commit:
KA-184 Installed Unity package

Commit:
KA-184 Implement Unity dependency resolver

Also, removed excess code and the paramless constructor of ItemController.

Commit:
KA-184 Generate a GUID for posted item

Commit:
KA-184 Add injectable GUID generator service

Commit:
KA-184 Add test that checks if a guid is generated in POST

Commit:
KA-184 Create V1 folder to simplify merge with task 1

Commit:
KA-184 Move ItemsController to V1 folder

Commit:
KA-184 Fix namespace in ItemsController

Commit:
KA-184 Merge in changes from task-1

Commit:
KA-184 Fix some minor issues in the tests, remove regions

Wrapped the NSubstitute calls in asserts

Commit:
KA-184 Move Repositories to own project

Commit:
KA-184 Add Contracts assembly

Commit:
KA-184 Move contracts to their own assembly

Commit:
KA-184 Add Utils project and move utils there

Commit:
KA-184 Move ListItemRepository

Commit:
KA-184 Rename assemblies

Commit:
KA-184 Fix using namespaces

Commit:
KA-184 Move the unity resolver back to Api project

Commit:
KA-184 Remove unused using directives

Commit:
KA-184 Implement the disposal of controller after tests

Commit:
KA-184 Use object initializer in test setup

Commit:
KA-184 Refactor Put to utilize repo.update

Commit:
KA-184 Make contracts internal

Commit:
KA-184 Remove the guid generator

Commit:
KA-184 Make contracts publci again, repo is now internal

Commit:
KA-184 Set the JSON serializer to ignore camel case

Commit:
KA-184 Remove the ID generator service

Commit:
KA-184 Name routes, pass to location response

Commit:
KA-184 Implement the unity bootstrapper for repo

Commit:
KA-184 Simplify the IRepository

The interface is no longer generic, the "Clear" method had been removed.

Commit:
KA-184 Decompose WebApiConfig

Commit:
KA-184 Rename WebApiconfig

Commit:
KA-184 Remove unused using directives

Commit:
KA-184 Move assemblies to Sources and Tests folders

Commit:
KA-184 Fix references and package paths

Commit:
KA-184 Fix names, remove obsolete comments

Commit:
KA-184 Remove Clear implementation from repo

Commit:
KA-184 Add routing to tests

Commit:
KA-184 Convert the bootstrapper extension to lambda

Commit:
KA-184 Move bootstrapper to its folder

Commit:
KA-184 Create IBootstrapper interface, rename implementation

Commit:
KA-184 Change register method to lambda

Commit:
KA-184 Use the newly created IBootstrapper

Commit:
KA-184 Change configs to internal

Commit:
KA-184 Make Repo methods fluent, remove key from Add

Commit:
KA-184 Create bootstrapper interface

Also, IRouteHelperConfig was created to prepare for future commits

Commit:
KA-184 Refactor repo bootstrapper to use interface

Commit:
KA-184 Implement RouteHelper

Route helper is used to get Url from corresponding route with given parameter.

Commit:
KA-184 Implement API bootstrapper

This bootstrapper registers RouteHelperConfig of this API as well as the function for obtaining current HttpRequestMessage which is used by the RouteHelper

Commit:
KA-184 Refactor controller routing, use RouteHelper

Commit:
KA-184 Mock RouteHelper in tests

Commit:
KA-184 Rename RegisterBootstrapper, squish code

Commit:
KA-184 Squish RegisterTypes into one lambda

Commit:
KA-184 Rename ItemsRouteHelperConfig

Commit:
KA-184 remove RouteHelperConfig ctor

Commit:
KA-184 Change methods to lambdas, remove comments

Commit:
KA-184 Fix expected ListItems in tests

Commit:
KA-184 Add ListApp.Api.Services assembly

Commit:
KA-184 Rename Services folder in Api to Bootstrapper

Commit:
KA-184 Move the RouteHelper to Services assembly

Commit:
KA-184 Fix references

Commit:
KA-184 Delete now returns deleted item

Commit:
KA-184 Refactor Delete tests to check returned item

vit-svoboda commented on ListApp/Sources/ListApp.Api/Controllers/V1/ItemsController.cs:
```diff
@@ -0,0 +1,55 @@
+    {
+        private readonly IRepository _repository;
+        private readonly IRouteHelper _routeHelper;
+        internal const string RouteName = "itemsRoute";
```
The constant should be above the fields. Perhaps also separated by an empty line.

vit-svoboda commented on ListApp/Sources/ListApp.Api/Controllers/V1/ItemsController.cs:
```diff
@@ -0,0 +1,55 @@
+            var deletedItem = await _repository.DeleteAsync(id);
+           
+            return Ok(deletedItem);
+        }
```
Put and Delete could be expression bodied functions, just like the Gets.

vit-svoboda commented on ListApp/Sources/ListApp.Api.Services/RouteHelper/IRouteHelper.cs:
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace ListApp.Api.Services.RouteHelper
```
Very nice, but should be in the ListApp.Contracts assembly.

vit-svoboda commented on ListApp/Sources/ListApp.Api.Services/RouteHelper/RouteHelperBootstrapper.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace ListApp.Api.Services.RouteHelper
+{
+    public class RouteHelperBootstrapper : IUnityContainerBootstrapper
```
The bootstrapper indeed handles just the RouteHelper right now but should be there for the whole assembly. So something like an ApiServicesBootstrapper would be a better name. And I'd place in the root of the project, or in the Bootstrapper folder, you have in the other projects.

milan009 replied:
Is this a good idea though? Imagine a situation when the Services assembly contains more classes than just RouteHelper (eg. an GuidGenerator). If there was only one bootstrapper for entire assembly, whenever using any service out of the assembly, all implementation of all services would register. What if wanted to use the provided RouteHelper implementation, but implement GuidGenerator service on my own? I believe that this exact situation was the reason why IRouteHelper interface should be moved to Contracts assembly.

vit-svoboda replied:
In that case, I'm guessing you should have an assembly for RouteHelper and an assembly for GuidGenerator. Otherwise, you'd bring to your application more than you need in the case you want your own GuidGenerator implementation. Of course, you can overwrite the registration in DI container, but that's not very nice, as you'd probably guess. Look at an assembly as a logical unit of high-level functionality you bring to your application. The bootstrapper is sort of an entry point to it.

vit-svoboda commented on ListApp/Sources/ListApp.Contracts/Interfaces/IRepository.cs:
```diff
@@ -0,0 +1,52 @@
+        /// A method that returns all keys from repository
+        /// </summary>
+        /// <returns>A collection of all keys</returns>
+        Task<IEnumerable<Guid>> GetKeysAsync();
```
I don't think this is a very useful method.

vit-svoboda commented on ListApp/Sources/ListApp.Contracts/Interfaces/IRepository.cs:
```diff
@@ -0,0 +1,52 @@
+        /// <summary>
+        /// A method that returns a specific item identified by given key.
+        /// </summary>
+        /// <param name="key"></param>
```
No param comment would be better than this, I guess.

vit-svoboda commented on ListApp/Sources/ListApp.Contracts/Interfaces/IRepository.cs:
```diff
@@ -0,0 +1,52 @@
+        /// </summary>
+        /// <param name="key"></param>
+        /// <returns>The item with given key, null if no such item exists.</returns>
+        Task<ListItem> GetAsync(Guid key);
```
This interface is specific to items that are identified by an ID. So the parameter is not a key, but an id or itemId.

vit-svoboda commented on ListApp/Sources/ListApp.Contracts/Interfaces/IRepository.cs:
```diff
@@ -0,0 +1,52 @@
+        /// </summary>
+        /// <param name="entity">The entity to store in the repository</param>
+        /// <exception cref="ArgumentNullException">if a null argument is passed</exception>
+        /// <exception cref="DuplicateKeyException">if an item with given key already exists</exception>
```
Saying the method will throw some kind of an exception here in an interface is rather controversial. However, the DuplicateKeyException is almost certainly wrong. Is it the one from System.Data.Linq? Will the implementation or implementations really throw that one?

vit-svoboda commented on ListApp/Sources/ListApp.Contracts/Interfaces/IRepository.cs:
```diff
@@ -0,0 +1,52 @@
+        /// <param name="entity">The entity to store in the repository</param>
+        /// <exception cref="ArgumentNullException">if a null argument is passed</exception>
+        /// <exception cref="DuplicateKeyException">if an item with given key already exists</exception>
+        Task<ListItem> AddAsync(ListItem entity);
```
The item you're adding is just an "item". The "entity" seems needlessly abstract.

vit-svoboda commented on ListApp/Sources/ListApp.Contracts/Interfaces/IRepository.cs:
```diff
@@ -0,0 +1,52 @@
+        /// </summary>
+        /// <param name="key">Key indetifyign the entity to update</param>
+        /// <param name="entity">The entity containing the updated data</param>
+        Task<ListItem> UpdateAsync(Guid key, ListItem entity);
```
The item already contains an ID. Now the repository has to decide which one is the right one, key or entity.Id.

vit-svoboda commented on ListApp/Sources/ListApp.Utils/ListApp.Utils.csproj:
```diff
@@ -0,0 +1,60 @@
+    <Reference Include="System.Xml" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="Constants.cs" />
```
The only purpose of this project is to store fake values you return from the repository right now. I think the unfinished repository implementation should make the values up on its own. Also, you don't want to share values between implementation (even though it's just a mock as of now) and the tests. So I'd suggest getting rid of this project.

vit-svoboda commented on ListApp/Tests/ListApp.Api.Tests/Extensions/EqualConstraintExtensions.cs:
```diff
@@ -1,6 +1,6 @@
+using ListApp.Contracts.Models;
 
 namespace ListApp.Api.Tests.Extensions
 {
```
Petr really doesn't like abbreviations, so I'd suggest the whole "constraint". As a bonus, it'll stop lighting up the spellchecker. What I'd suggest as well if you don't have one yet.

milan009 replied:
Thanks for the suggestion, it has never really occurred to me that there might be a spellchecker extension for VS. D'oh. Removed the abbr. in #5ba514c.

vit-svoboda commented on ListApp/Tests/ListApp.Api.Tests/ItemControllerTests.cs:
```diff
@@ -21,17 +25,44 @@ public class ItemsControllerTests
+            _itemsRepository.DeleteAsync(Arg.Any<Guid>())
+                .Returns(Constants.MockListItems.ElementAt(0));
+            _itemsRepository.UpdateAsync(Arg.Any<Guid>(), Arg.Any<ListItem>())
+                .Returns(Constants.MockListItems.ElementAt(0));
```
You don't need to mock all of the methods here. The test methods can mock the method(s) they need with their own expected results.

vit-svoboda commented on ListApp/Tests/ListApp.Api.Tests/ItemControllerTests.cs:
```diff
@@ -21,17 +25,44 @@ public class ItemsControllerTests
+        
         [Test]
-        public async Task Get_NoId_ResponseIsOfCorrectTypeAndReturnsDefaultItems()
+        public async Task Get_NoId_ResponseIsOfCorrectTypeAndReturnsDefaultItemsAndCallsRepoGetAsyncMethodOnce()
```
Get_NoId_TLDR

vit-svoboda commented on ListApp/Tests/ListApp.Api.Tests/ItemControllerTests.cs:
```diff
@@ -62,60 +94,69 @@ public void SetUp()
             var receivedResponse = await _itemsController.GetAsync(Guid.NewGuid());
             var responseMessage = await receivedResponse.ExecuteAsync(CancellationToken.None);
 
+            Assert.DoesNotThrowAsync(() => _itemsRepository.Received(1).GetAsync(Arg.Any<Guid>()));
```
This DoesNotThrowAsync call is pointless, the Received throws nice enough exception that fails the test as well.

vit-svoboda commented on ListApp/Tests/ListApp.Api.Tests/ItemControllerTests.cs:
```diff
@@ -62,60 +94,69 @@ public void SetUp()
             var responseMessage = await receivedResponse.ExecuteAsync(CancellationToken.None);
 
+            Assert.DoesNotThrowAsync(() => _itemsRepository.Received(1).GetAsync(Arg.Any<Guid>()));
             Assert.AreEqual(expectedResponseCode, responseMessage.StatusCode);
```
Assert.That is preferred in all cases.

vit-svoboda commented on ListApp/Tests/ListApp.Api.Tests/ItemControllerTests.cs:
```diff
@@ -62,60 +94,69 @@ public void SetUp()
 
+            Assert.DoesNotThrowAsync(() => _itemsRepository.Received(1).DeleteAsync(PostedItemGuid));
             Assert.AreEqual(expectedResponseCode, responseMessage.StatusCode);
+            Assert.IsTrue(responseMessage.TryGetContentValue(out ListItem responseItem));
```
You don't need to be this defensive in the tests. Just get the content value in the Act part of the test and then check it in the last Assert.

vit-svoboda reviewed (changes requested):
Nicely done. The comments bellow mostly just point out things that could be named more accurately.

Commit:
KA-184 Move the constant to the top of the controller

Commit:
KA-184 Change Put and Delete to expression bodied

Commit:
KA-184 Move IRouteHelper to contracts

Commit:
KA-184 Refactor the IRepository interface

Renamed parameters to make them more self-explanatory, Update no longer requires an ID, as a valid ID should be included in the passed item.

Corresponding changes to the implementation an the tests have been made to reflect this refactoring.

Commit:
KA-184 Got rid of that one last abbreviation

RIP

Commit:
KA-184 Removed the GetKeys() method

Commit:
KA-184 Fix IRouteHelper namespace

Commit:
KA-184 Remove the Utils project

Commit:
KA-184 Remove unnecessary assert in tests

Additionally, Assert.Equals was replaced by Assert.That

Commit:
KA-184 Attempt to rename tests to more sensible names

vit-svoboda commented on ListApp/Tests/ListApp.Api.Tests/ListApp.Api.Tests.csproj:
```diff
@@ -88,10 +88,6 @@
       <Project>{f895dfd4-2ab2-4935-9da8-8d2a5c3d91fa}</Project>
```
Now you've moved the IRouteHelper to contract, tests don't need to reference ListApp.Api.Services anymore.

vit-svoboda commented on ListApp/Tests/ListApp.Api.Tests/ItemControllerTests.cs:
```diff
@@ -62,100 +46,107 @@ public void TearDown()
         
         [Test]
-        public async Task Get_NoId_ResponseIsOfCorrectTypeAndReturnsDefaultItemsAndCallsRepoGetAsyncMethodOnce()
+        public async Task Get_NoId_ReturnsExpectedResponse()
```
ReturnsOkResponse might be a bit more apt while keeping the name short. An OK response is what you expect here.

vit-svoboda reviewed (changes requested):
Very nice, there is literally just a couple things you can make better. Plus that bootstrapper discussion from earlier.

Commit:
KA-184 Remove unused references

Commit:
KA-184 Rename test to be more specific

Commit:
KA-184 Rename the boostrapper in Services assembly

Commit:
KA-184 Move the ServiceApiBootstrapper to its own folder

vit-svoboda commented on ListApp/Sources/ListApp.Api.Services/Bootstrapper/ServicesApiBootstrapper.cs:
```diff
@@ -3,7 +3,7 @@
 
 namespace ListApp.Api.Services.RouteHelper
```
Now the namespace does not correspond the folder structure. There's a ReSharper refactoring to fix that.

vit-svoboda reviewed (changes requested):
I must have missed it last time, but there are some unused using statements in the ItemsController. Also, I'd rename the ListItemRepositoryBootstrapper to RepositoriesBootstrapper, just like you renamed the ServicesApiBootstrapper.

Oh and one last thing - config files are generally useful only for entry point assemblies, i.e. executable files, test libraries and web applications. So I'm quite positive you can delete the app.config in ListApp.Repositories.

Commit:
KA-184 Remove extra using directives, fix namespaces

Commit:
KA-184 Remove the app.config in repositories assembly

milan009 commented:
Good points, thanks. Should be fixed now

Commit:
KA-184 Fix indentation

Commit:
KA-184 Rename private field and make it readonly

vit-svoboda reviewed (approved):
Awesome. I noticed you don't `await` the `Received` checks of async methods. It doesn't really make any difference, but some static analysis tools (e.g. ReSharper) might report those lines as suspicious. So adding the await statements could make the code a bit easier to maintain. Just so you know ;)
